### PR TITLE
Migrate all auths to use the auth interface

### PR DESCRIPTION
**Pull request by harryi3t**

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -129,7 +144,7 @@
         "no-restricted-globals": "error",
         "no-shadow": "off",
         "no-shadow-restricted-names": "error",
-        "no-undef": "off",
+        "no-undef": 2,
```
Comment (kamalaknn) [inline .eslintrc]: @harryi3t Should this be part of a separate PR?

File: lib/authorizer/basic.js [hunk 2]
```diff
@@ -39,14 +39,12 @@ module.exports = {
...
-     * @param {RequestAuth} auth
+     * @param {AuthInterface} auth
      * @param {Request} request
      * @param done
      */
     sign: function (auth, request, done) {
-        var params = auth.parameters().toObject();
-
-        if (!params) { return done(); } // Nothing to do if no parameters are present.
+        var params = auth.get(['username', 'password']);
```
Comment (kamalaknn) [inline lib/authorizer/basic.js]: Why not `params.get('username')` and `params.get('password')` here? It is just two variables right?

File: lib/authorizer/digest.js [hunk 3]
```diff
@@ -91,32 +89,21 @@ module.exports = {
...
-     * @param {String} params.nonce
-     * @param {String} params.nonceCount
-     * @param {String} params.clientNonce
-     * @param {String} params.opaque
-     * @param {String} params.qop
-     * @param {String} params.uri
+     * @param {AuthInterface} auth
      * @returns {String}
      */
-    computeHeader: function (params) {
```
Comment (kamalaknn) [inline lib/authorizer/digest.js]: This is the only Auth where `computeHeader` takes `AuthInterface` instead of Object. Should we stick to one?

File: lib/authorizer/hawk.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ var _ = require('lodash'),
  * @param {Number} length
  */
 function randomString (length) {
-    length = length || 6;
+    length = length || 8;
```
Comment (kamalaknn) [inline lib/authorizer/hawk.js]: Can you move this out of this PR?

File: lib/authorizer/hawk.js [hunk 5]
```diff
@@ -39,26 +39,22 @@ module.exports = {
...
+     * @param {AuthInterface} auth
      * @param done
      */
     pre: function (auth, done) {
-        var authVarList = auth.parameters(),
-            params = {};
-
-        params.nonce = params.nonce || randomString(6);
-        params.timestamp = _.parseInt(params.timestamp) ||
+        !auth.get('nonce') && auth.set('nonce', randomString(8));
```
Comment (kamalaknn) [inline lib/authorizer/hawk.js]: Here too.

File: lib/authorizer/digest.js [hunk 6]
```diff
@@ -178,23 +165,23 @@ module.exports = {
...
             !params.username ||
             !params.realm) { return done(); } // Nothing to do if no parameters are present.
 
         request.removeHeader('Authorization', {ignoreCase: true});
 
-        params.method = request.method;
-        params.uri = request.url.getPathWithQuery();
-        header = this.computeHeader(params);
+        auth.set('method', request.method);
+        auth.set('uri', request.url.getPathWithQuery());
```
Comment (kamalaknn) [inline lib/authorizer/digest.js]: Let us not save URI and method to auth.

File: lib/authorizer/hawk.js [hunk 7]
```diff
@@ -88,15 +84,24 @@ module.exports = {
...
+                'algorithm',
+                'nonce',
+                'timestamp',
+                'extraData',
+                'app',
+                'delegation',
+                'user'
+            ]);
+        if (!params.authId || !params.authKey) {
             return done(); // Nothing to do if no parameters are present.
```
Comment (kamalaknn) [inline lib/authorizer/hawk.js]: I'm okay with the change, but the comment is inconsistent now.

File: lib/authorizer/index.js [hunk 8]
```diff
@@ -104,14 +105,16 @@ authorizeRequest = function (request, done) {
 
     var clonedReq = new sdk.Request(request.toJSON()),
         auth = clonedReq.auth,
+        authInterface = createAuthInterface(auth),
         authVarlist = auth.parameters(), // @todo: wrap in auth interface
         handler;
 
     if (authVarlist) {
```
Comment (kamalaknn) [inline lib/authorizer/index.js]: Remove.

File: lib/authorizer/index.js [hunk 9]
```diff
@@ -104,14 +105,16 @@ authorizeRequest = function (request, done) {
...
         handler = AuthLoader.getHandler(auth.type);
     }
 
-    handler && handler.sign(auth, clonedReq, function () { done(null, clonedReq); });
+    handler && handler.sign(authInterface, clonedReq, function () { return done(null, clonedReq); });
+    // @todo what happens if handler is not found? why there is no return here??
```
Comment (kamalaknn) [inline lib/authorizer/index.js]: Maybe let's add `if (!handler) { done('runtime~authorizeRequest: could not find handler for auth type' + auth.type); } `?

File: lib/authorizer/ntlm.js [hunk 10]
```diff
@@ -33,28 +33,24 @@ module.exports = {
...
      *
-     * @param {RequestAuth} auth
+     * @param {AuthInterface} auth
      * @param {Response} response
      * @param done
      */
     post: function (auth, response, done) {
-        var authVarList = auth.parameters(),
-            state = authVarList.toObject().state,
+        var state = auth.get('state'),
```
Comment (kamalaknn) [inline lib/authorizer/ntlm.js]:
```
domain = auth.get('domain'),
workstation = auth.get('workstation'),
username = auth.get('username'),
passworkd = auth.get('password');
```
see Lines 73, 74, 101, 102

File: lib/authorizer/oauth1.js [hunk 11]
```diff
@@ -112,18 +112,31 @@ module.exports = {
...
+                'addEmptyParamsToSign'
+            ]),
             signatureParams,
             header;
 
-        if (!helperParams ||
-            !helperParams.consumerKey ||
-            !helperParams.consumerSecret) { return done(); } // Nothing to do if no parameters are present.
+        if (!params.consumerKey || !params.consumerSecret) {
+            return done(); // Nothing to do if no parameters are present.
```
Comment (kamalaknn) [inline lib/authorizer/oauth1.js]: Comment needs to be updated.

File: lib/runner/request-helpers-postsend.js [hunk 12]
```diff
@@ -17,10 +18,11 @@ module.exports = [
         if (!util.isInteractiveForAuth(run.options, context.auth.type)) { return done(); }
 
         var auth = context.auth,
-            authorizer = AuthLoader.getHandler(auth.type);
+            authorizer = AuthLoader.getHandler(auth.type),
```
Comment (kamalaknn) [inline lib/runner/request-helpers-postsend.js]: can you rename this to `authHandler`? I missed this spot.

File: lib/runner/request-helpers-presend.js [hunk 13]
```diff
@@ -1,6 +1,7 @@
 var _ = require('lodash'),
     async = require('async'),
     sdk = require('postman-collection'),
+    createAuthInterface = require('../authorizer/auth-interface'),
```
Comment (kamalaknn) [inline lib/runner/request-helpers-presend.js]: New line above?

File: lib/runner/request-helpers-presend.js [hunk 14]
```diff
@@ -41,19 +42,23 @@ module.exports = [
...
 
         // bail out if there is not matching auth handler for the type
         if (!authHandler) { return done(new Error('runtime: no handler for auth')); }
 
+        authInterface = createAuthInterface(auth);
+
         // sign and bail out if interactive mode is disabled
         if (!util.isInteractiveForAuth(run.options, authType)) { return authSignHook(); }
 
+
```
Comment (kamalaknn) [inline lib/runner/request-helpers-presend.js]: Extra newline.

File: test/unit/auth-handlers.test.js [hunk 15]
```diff
@@ -61,13 +62,15 @@ describe('Auth Handler:', function () {
             var rawBasicReq = _.cloneDeep(rawRequests.basic),
                 request,
                 auth,
+                authInterface,
                 handler;
 
             rawBasicReq.auth.basic = {username: 'foo'}; // no password present
             request = new Request(rawBasicReq);
             auth = request.auth;
+            authInterface = createAuthInterface(auth);
```
Comment (kamalaknn) [inline test/unit/auth-handlers.test.js]: Let's do `authInterface = createAuthInterface(request.auth);` everywhere.

File: lib/authorizer/digest.js [hunk 16]
```diff
@@ -171,17 +182,30 @@ module.exports = {
...
+                'nonce',
+                'nonceCount',
+                'clientNonce',
+                'opaque',
+                'qop',
+                'uri'
+            ]);
+
+        if (!params.username || !params.realm) {
+            return done(); // Nothing to do if no parameters are present.
```
Comment (kamalaknn) [inline lib/authorizer/digest.js]: Update comment.

File: lib/authorizer/index.js [hunk 17]
```diff
@@ -106,15 +106,12 @@ authorizeRequest = function (request, done) {
...
         auth = clonedReq.auth,
         authInterface = createAuthInterface(auth),
-        authVarlist = auth.parameters(), // @todo: wrap in auth interface
-        handler;
-
-    if (authVarlist) {
         handler = AuthLoader.getHandler(auth.type);
-    }
 
     handler && handler.sign(authInterface, clonedReq, function () { return done(null, clonedReq); });
```
Comment (kamalaknn) [inline lib/authorizer/index.js]: Change to a simple if else block.

File: lib/authorizer/basic.js (hunk 2 above)
Comment (shamasis) [reply lib/authorizer/basic.js]: Agree

File: lib/runner/request-helpers-presend.js [hunk 18]
```diff
@@ -41,32 +43,35 @@ module.exports = [
             authType = auth.type,
             authHandler = AuthLoader.getHandler(authType),
             authPreHook,
+            authInterface,
             authSignHook = function () {
                 try {
-                    authHandler.sign(auth, context.item.request, function (err) { done(err); });
+                    authHandler.sign(authInterface, context.item.request, function (err) { done(err); });
```
Comment (shamasis) [inline lib/runner/request-helpers-presend.js]: Sign is in a try block while pre post and init are not. That's okay for now?

File: lib/authorizer/index.js [hunk 19]
```diff
@@ -104,14 +105,15 @@ authorizeRequest = function (request, done) {
...
+        authInterface = createAuthInterface(auth),
         handler = AuthLoader.getHandler(auth.type);
-    }
 
-    handler && handler.sign(auth, clonedReq, function () { done(null, clonedReq); });
+    if (handler) {
+        handler.sign(authInterface, clonedReq, function () { return done(null, clonedReq); });
+    }
+    else {
+        return done('runtime~authorizeRequest: could not find handler for auth type' + auth.type);
```
Comment (kunagpal) [inline lib/authorizer/index.js]:
```js
done(new Error('...'));
```

File: lib/runner/request-helpers-presend.js (hunk 18 above)
Comment (kamalaknn) [reply lib/runner/request-helpers-presend.js]: The reason, why `sign` hook in a `try catch` is because

1. In the old implementation sign was the only synchronous API
2. We are using third party libraries mostly in sign hooks, which may not strictly adhere to sending error objects in callbacks.

Going back, we use third-party in other hooks as well.

Should we move other callbacks now, or wait till custom auth comes in?

File: test/unit/authorizer-request.test.js [hunk 20]
```diff
@@ -18,4 +20,30 @@ describe('.authorizeRequest (Static function)', function () {
...
+
+    it('must return error when handler is not found for a given auth', function () {
+        var fakeAuth = {
+                type: 'fake',
+                fake: {
+                    user: 'user',
+                    pass: 'pass'
+                }
+            },
+            request = new Request(_.assign(rawRequests.basic, {auth: fakeAuth}));
```
Comment (kamalaknn) [inline test/unit/authorizer-request.test.js]: `_.assing({}, awRequests.basic, {auth: fakeAuth})`

File: lib/runner/request-helpers-presend.js (hunk 18 above)
Comment (harryi3t) [reply lib/runner/request-helpers-presend.js]: NTLM is already using 3rd party module in its post hook, so we should do this.
@shamasis should I do this in a separate PR?
